File: typescript/robot-simulator/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message: string) {
    super()
    this.message = message || 'Invalid Input'
  }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

export class Robot {
    public coords: Coordinates = [0, 0]
    public direction: Direction = 'north'

  get bearing(): Direction {
    return this.direction
  }

  get coordinates(): Coordinates {
    return this.coords
  }

  place(args: { x: number; y: number; direction: string }) {
    this.coords = [args.x, args.y]
    if (['north', 'east', 'south', 'west'].includes(args.direction)) {
      this.direction = args.direction as Direction
    } else {
        throw new InvalidInputError('Invalid Input')
    }
  }

    turnRight() {
        const directions: Direction[] = ['north', 'east', 'south', 'west']
        const index = directions.indexOf(this.direction)
        this.direction = directions[(index + 1) % 4]
    }

    turnLeft() {
        const directions: Direction[] = ['north', 'west', 'south', 'east']
        const index = directions.indexOf(this.direction)
        this.direction = directions[(index + 1) % 4]
    }

    advance() {
        switch (this.direction) {
            case 'north':
                this.coords[1]++
                break
            case 'east':
                this.coords[0]++
                break
            case 'south':
                this.coords[1]--
                break
            case 'west':
                this.coords[0]--
                break
        }
    }

  evaluate(instructions: string) {
    instructions.split('').forEach((instruction) => {
      switch (instruction) {
        case 'L':
          this.turnLeft()
          break
        case 'R':
          this.turnRight()
          break
        case 'A':
          this.advance()
          break
        default:
          throw new InvalidInputError('Invalid Input')
      }
    })
  }
}

File: python/robot-simulator/robot_simulator.py
# Globals for the directions
# Change the values as you see fit
EAST = (1, 0)
NORTH = (0, 1)
WEST = (-1, 0)
SOUTH = (0, -1)
DIRECTIONS = [NORTH, EAST, SOUTH, WEST]


class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.x = x_pos
        self.y = y_pos

    @property
    def coordinates(self):
        return (self.x, self.y)

    def turn_right(self):
        index = DIRECTIONS.index(self.direction)
        self.direction = DIRECTIONS[(index + 1) % 4]

    def turn_left(self):
        index = DIRECTIONS.index(self.direction)
        self.direction = DIRECTIONS[(index - 1) % 4]

    def advance(self):
        self.x += self.direction[0]
        self.y += self.direction[1]

    def move(self, moves=""):
        entries = list(moves)
        for entry in entries:
            match entry:
                case 'R':
                    self.turn_right()
                case 'L':
                    self.turn_left()
                case 'A':
                    self.advance()
